zenoh-ext/src/net/querying_subscriber.rs

```diff
@@ -0,0 +1,483 @@
+        self.info.mode = SubMode::Pull;
+        self.info.period = period;
+        self
+    }
+    /// Change the resource key to be used for queries.
+    pub fn query_reskey(mut self, query_reskey: ResKey) -> Self {
+        self.query_reskey = query_reskey;
+        self
+    }
+
+    /// Change the predicate to be used for queries.
+    pub fn query_predicate(mut self, query_predicate: String) -> Self {
+        self.query_predicate = query_predicate;
+        self
+    }
+
+    /// Change the target to be used for queries.
+    pub fn query_target(mut self, query_target: QueryTarget) -> Self {
+        self.query_target = query_target;
+        self
+    }
+
+    /// Change the consolidation mode to be used for queries.
+    pub fn query_consolidation(mut self, query_consolidation: QueryConsolidation) -> Self {
+        self.query_consolidation = query_consolidation;
+        self
+    }
+}
+
+impl<'a> Future for QueryingSubscriberBuilder<'a> {
+    type Output = ZResult<QueryingSubscriber<'a>>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Poll::Ready(QueryingSubscriber::new(Pin::into_inner(self).clone()))
+    }
+}
+
+impl<'a> ZFuture<ZResult<QueryingSubscriber<'a>>> for QueryingSubscriberBuilder<'a> {
+    fn wait(self) -> ZResult<QueryingSubscriber<'a>> {
+        QueryingSubscriber::new(self)
+    }
+}
+
+pub struct QueryingSubscriber<'a> {
+    conf: QueryingSubscriberBuilder<'a>,
+    subscriber: Subscriber<'a>,
+    receiver: QueryingSubscriberReceiver,
+}
+
+impl QueryingSubscriber<'_> {
+    fn new(conf: QueryingSubscriberBuilder<'_>) -> ZResult<QueryingSubscriber<'_>> {
+        // declare subscriber at first
+        let mut subscriber = conf
+            .session
+            .declare_subscriber(&conf.sub_reskey, &conf.info)
+            .wait()?;
+
+        let receiver = QueryingSubscriberReceiver::new(subscriber.receiver().clone());
+
+        let mut query_subscriber = QueryingSubscriber {
+            conf,
+            subscriber,
+            receiver,
+        };
+
+        // start query
+        query_subscriber.query()?;
+
+        Ok(query_subscriber)
+    }
+
+    /// Undeclare this QueryingSubscriber
+    #[inline]
+    pub fn undeclare(self) -> ZResolvedFuture<ZResult<()>> {
+        self.subscriber.undeclare()
+    }
+
+    /// Return the QueryingSubscriberReceiver associated to this subscriber.
+    #[inline]
+    pub fn receiver(&mut self) -> &mut QueryingSubscriberReceiver {
+        &mut self.receiver
+    }
+
+    /// Issue a new query using the configured resource key and predicate.
+    pub fn query(&mut self) -> ZResult<()> {
+        self.query_on(
+            &self.conf.query_reskey.clone(),
+            &self.conf.query_predicate.clone(),
+            self.conf.query_target.clone(),
+            self.conf.query_consolidation.clone(),
+        )
+    }
+
+    /// Issue a new query on the specified resource key and predicate.
+    pub fn query_on(
```

Comment:
Shouldn't this operation be async ? It issues a query and waits for it, shouldn't it await?

Reply:
This doesn't wait for the replies. But to be coherent with zenoh::net::Session::query() it could return a ZFuture anyway. (Basically, almost everything return a ZFuture now in zenoh APIs)

Reply:
No, this operation doesn't wait for the query. It triggers the query and store the resulting replies receiver in its InnerState. Then, the reception operations (poll_next/recv/try_recv...) are in charge to treat the query replies, adding those to the merge_queue, and then at query completion to sort/filter/forward the merge_queue to the user.

The result is that the wait for query completion is actually made in the reception operations (or not in the case of `poll_next()` and `try_recv()` that return respectively `Poll::Pending` and `TryRecvError::Empty` if the query is not complete)

Reply:
> This doesn't wait for the replies. But to be coherent with zenoh::net::Session::query() it could return a ZFuture anyway. (Basically, almost everything return a ZFuture now in zenoh APIs)

OK. I'll add this as well as management of a list of queries in progress, instead of returning an error if one is already in progress.

Reply:
Done in c253b7a.